snuba/admin/static/clickhouse_queries/index.tsx

```diff
@@ -1,20 +1,137 @@
 import { QueryResult } from "../components/query_display/types";
 
+type PredefinedQuery = {
+  name?: string;
+  sql?: string;
+  description?: string;
+  selected?: boolean;
+};
+
 function ClickhouseQueries(props: { api: Client }) {
+  const [predefinedQuery, setPredefinedQuery] = useState<PredefinedQuery>({});
+  const [predefinedQueryOptions, setPredefinedQueryOptions] = useState<
+    PredefinedQuery[]
+  >([]);
+
+  useEffect(() => {
+    props.api.getPredefinedQueryOptions().then((res) => {
+      res.forEach(
+        (queryOption) => (queryOption.sql = formatSQL(queryOption.sql))
+      );
+      setPredefinedQueryOptions(res);
+    });
+  }, []);
+
   function tablePopulator(queryResult: QueryResult) {
     return (
       <Table headerData={queryResult.column_names} rowData={queryResult.rows} />
     );
   }
-  return QueryDisplay({
-    api: props.api,
-    endpoint: "run_clickhouse_system_query",
-    resultDataPopulator: tablePopulator,
-  });
+
+  function updatePredefinedQuery(queryName: string) {
+    const selectedQuery = predefinedQueryOptions.find(
+      (query) => query.name === queryName
+    );
+    setPredefinedQuery(() => {
+      return {
+        ...selectedQuery,
+        selected: false,
+      };
+    });
+  }
+
+  function selectPredefinedQuery() {
+    setPredefinedQuery((prevQuery) => {
+      return {
+        ...prevQuery,
+        selected: true,
+      };
+    });
+  }
+
+  function clearPredefinedQuery() {
+    setPredefinedQuery(() => {
+      return {};
+    });
+  }
+
+  function formatSQL(sql: string) {
+    const formatted = sql
+      .split("\n")
+      .map((line) => line.substring(4, line.length))
+      .join("\n");
+    return formatted.trim();
+  }
+
+  return (
+    <div>
+      <div>
+        <form>
+          <select
+            value={predefinedQuery.name || ""}
+            onChange={(evt) => updatePredefinedQuery(evt.target.value)}
+            style={selectStyle}
+          >
+            <option disabled value="">
+              Select a predefined query
+            </option>
+            {predefinedQueryOptions.map((option: PredefinedQuery) => (
+              <option key={option.name} value={option.name}>
+                {option.name}
+              </option>
+            ))}
+          </select>
+        </form>
+        {predefinedQuery?.sql && (
+          <div>
+            {
+              <div
+                style={{
+                  fontSize: 14,
+                  marginBottom: 5,
+                  width: "50vw",
+                }}
+              >
+                <p
+                  dangerouslySetInnerHTML={{
+                    __html: predefinedQuery.description || "",
+                  }}
+                ></p>
+              </div>
+            }
+            <textarea
+              readOnly={true}
+              value={predefinedQuery?.sql}
+              spellCheck={false}
+              style={{ width: "50vw", height: 100 }}
+            />
```

Comment:
How come we are adding 2 text boxes with the query and extra step to copy it over to the main input? Why not just populate the main input area to begin with?

Reply:
So I think it's nice to be able to click between the precanned queries without them messing up the current query you have. Im thinking of these precanned queries as basically as reference of what you could do. Like you may not know just by the name if it's the _exact_ query you want. I like being able to see what the sql is before I decide I want to use it. 

open to thoughts here if it ends up being annoying for folks. much easier to take away later than to add back in.

Reply:
I agree that it is nice to keep the two separate. That way I can do something like copy the query down into the edit box, and keep the precanned version visible so I know how to populate the parameterized values.

I think the Clear button is weird though, since it's up with the precanned queries but erases what is in the editing textbox. I'm not sure that button is useful at all.

Reply:
I think the additional text box makes the whole system queries page (both when you are using canned queries and without) more visually complicated.

Some thoughts:
- There are now two input boxes on the page - imo the wrong one is emphasised at the top.
- It looks like you can and should edit the predefined queries text box but you actually can't.
- There's a lot more buttons and clicks involved when you are using a predefined query as you need to send it through 2 text boxes to run it instead of just selecting it and hitting Execute which seems much more straightforward to me.

Ultimately I think the double textbox design may be optimizing for what could be a pretty niche use case (wanting to see multiple queries together both predefined and self entered) at the detriment of the regular experience where you just want to type a query or select one predefined query only. I am not opposed to this design if many people feel that they want to use the tool with this workflow, but I don't see myself doing that.

i think that keeping UIs simple is pretty hard even though it looks easy, and I disagree with the notion that we should add just in case because taking away is easier than adding things.

Reply:
@evanh the clear shouldn't change whats in the edited text box, should just clear the top part so its not in the way if you've moved on from using the precanned queries. so if it's doing that it wasn't intended

Reply:
Ah OK. I'll have to test again to see if that's what was happening. I would maybe change that from Clear -> Hide or maybe Collapse so it's a little more clear what's happening. Obviously a very small nit though.

Reply:
> Ultimately I think the double textbox design may be optimizing for what could be a pretty niche use case (wanting to see multiple queries together both predefined and self entered) at the detriment of the regular experience where you just want to type a query or select one predefined query only. I am not opposed to this design if many people feel that they want to use the tool with this workflow, but I don't see myself doing that.


@lynnagara I think this is a good point - I def built this out in a way that I felt was useful for me. Maybe if I can clarify the goals I was trying to solve by doing it this way, folks may have other ideas on how to achieve that. 

1. Since we won't be able to have every possible query we'd want, my guess was that we'd have a lot of canned queries that people would want to edit. I wanted to make it possible to quickly see the original query without it overriding what ever you had edited, incase you had removed something you wanted back in. It's possible you could get this from the result history, but I find that very hard to use in its current state.
2. We talked about having this page have more info on what options you have to even query by in clickhouse without going to the system.tables documentation. I felt the easiest way to do this was to have a bunch of canned queries that utilized those options (different columns from different tables) to use as a reference. Again, if you are in the middle of building a query, I think I would find it annoying showing a pre-canned query overrode what i was currently working on. 

I guess the overall goal I was trying to do was to make this useful as a reference while building your own queries and not just a plug in and execute functionality.